File: samples/CameraAccess/CameraAccess/Display/MusicPlayer.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

//
// MusicPlayer.swift
//
// Plays a music URL (e.g. an Eleven Labs track returned by OpenClaw) via AVPlayer.
// The DAT SDK has no glasses audio-playback API, so playback uses iOS audio, which
// routes to the Ray-Ban Meta glasses over Bluetooth (A2DP) when they are connected
// as the active output. MusicPlayerWidget is the play/pause card shown in the FOV.
//

import AVFoundation
import Combine
import SwiftUI

@MainActor
final class MusicPlayer: ObservableObject {
  static let shared = MusicPlayer()

  @Published private(set) var isPlaying = false
  @Published private(set) var currentURL: String?

  private var player: AVPlayer?
  private var endObserver: NSObjectProtocol?

  private init() {}

  /// True if the given URL is the current track and it is playing.
  func isPlaying(url: String) -> Bool {
    isPlaying && currentURL == url
  }

  /// Starts (or resumes) playback of `url`. Audio routes to the active output,
  /// including Ray-Ban Meta glasses when connected over Bluetooth.
  func play(url: String) {
    guard let parsed = URL(string: url) else { return }
    configureSession()
    if currentURL != url || player == nil {
      loadItem(parsed, url: url)
    }
    player?.play()
    isPlaying = true
  }

  func pause() {
    player?.pause()
    isPlaying = false
  }

  func toggle(url: String) {
    if isPlaying(url: url) {
      pause()
    } else {
      play(url: url)
    }
  }

  func stop() {
    player?.pause()
    player = nil
    isPlaying = false
    currentURL = nil
  }

  // MARK: - Private

  private func loadItem(_ parsed: URL, url: String) {
    if let endObserver {
      NotificationCenter.default.removeObserver(endObserver)
      self.endObserver = nil
    }
    let item = AVPlayerItem(url: parsed)
    player = AVPlayer(playerItem: item)
    currentURL = url
    endObserver = NotificationCenter.default.addObserver(
      forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main
    ) { [weak self] _ in
      Task { @MainActor in self?.isPlaying = false }
    }
  }

  private func configureSession() {
    let session = AVAudioSession.sharedInstance()
    try? session.setCategory(.playback, options: [.allowBluetoothA2DP, .mixWithOthers])
    try? session.setActive(true)
  }
}

/// Native music-player card: a play/pause toggle + the track title.
struct MusicPlayerWidget: View {
  let url: String
  let title: String
  @ObservedObject private var player = MusicPlayer.shared

  var body: some View {
    HStack(spacing: 14) {
      Button {
        player.toggle(url: url)
      } label: {
        Image(systemName: player.isPlaying(url: url) ? "pause.circle.fill" : "play.circle.fill")
          .font(.system(size: 44))
          .foregroundColor(.white)
      }
      .buttonStyle(.plain)

      VStack(alignment: .leading, spacing: 2) {
        Text(title)
          .font(.system(size: 16, weight: .semibold))
          .foregroundColor(.white)
        Text(player.isPlaying(url: url) ? "Playing on your glasses…" : "Tap to play")
          .font(.system(size: 12))
          .foregroundColor(.white.opacity(0.6))
      }
      Spacer(minLength: 0)
    }
    .padding(16)
    .frame(maxWidth: .infinity, alignment: .leading)
    .displayCardStyle()
  }
}

File: samples/CameraAccessAndroid/app/src/main/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/display/MusicPlayer.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.meta.wearable.dat.externalsampleapps.cameraaccess.display

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager as SystemAudioManager
import android.media.MediaPlayer
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Plays a music or podcast URL (e.g. an Eleven Labs track returned by OpenClaw) via
 * [MediaPlayer]. The DAT SDK has no glasses audio-playback API, so playback uses
 * Android audio, which routes to Ray-Ban Meta glasses over Bluetooth when connected.
 */
object MusicPlayer {
    private const val TAG = "MusicPlayer"

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _currentUrl = MutableStateFlow<String?>(null)
    val currentUrl: StateFlow<String?> = _currentUrl.asStateFlow()

    private var appContext: Context? = null
    private var mediaPlayer: MediaPlayer? = null
    private var audioFocusRequest: AudioFocusRequest? = null

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    fun isPlaying(url: String): Boolean = _isPlaying.value && _currentUrl.value == url

    fun play(url: String) {
        val context = appContext ?: return
        if (_currentUrl.value != url || mediaPlayer == null) {
            releasePlayer()
            _currentUrl.value = url
            mediaPlayer =
                MediaPlayer().apply {
                    setAudioAttributes(
                        AudioAttributes.Builder()
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                            .build(),
                    )
                    setOnPreparedListener {
                        requestAudioFocus(context)
                        start()
                        _isPlaying.value = true
                    }
                    setOnCompletionListener {
                        _isPlaying.value = false
                        abandonAudioFocus(context)
                    }
                    setOnErrorListener { _, what, extra ->
                        Log.w(TAG, "Playback error what=$what extra=$extra")
                        _isPlaying.value = false
                        abandonAudioFocus(context)
                        true
                    }
                    try {
                        setDataSource(url)
                        prepareAsync()
                    } catch (e: Exception) {
                        Log.w(TAG, "Failed to load $url", e)
                        releasePlayer()
                    }
                }
        } else {
            requestAudioFocus(context)
            mediaPlayer?.start()
            _isPlaying.value = true
        }
    }

    fun pause() {
        mediaPlayer?.pause()
        _isPlaying.value = false
        appContext?.let { abandonAudioFocus(it) }
    }

    fun toggle(url: String) {
        if (isPlaying(url)) pause() else play(url)
    }

    fun stop() {
        releasePlayer()
        _isPlaying.value = false
        _currentUrl.value = null
    }

    private fun releasePlayer() {
        appContext?.let { abandonAudioFocus(it) }
        mediaPlayer?.reset()
        mediaPlayer?.release()
        mediaPlayer = null
    }

    private fun requestAudioFocus(context: Context) {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as SystemAudioManager
        val request =
            AudioFocusRequest.Builder(SystemAudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                        .build(),
                )
                .build()
        audioFocusRequest = request
        audioManager.requestAudioFocus(request)
    }

    private fun abandonAudioFocus(context: Context) {
        val request = audioFocusRequest ?: return
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as SystemAudioManager
        audioManager.abandonAudioFocusRequest(request)
        audioFocusRequest = null
    }
}
